sdk/storage_datalake/src/shared_key_authorization_policy.rs

```diff
@@ -0,0 +1,244 @@
+use azure_core::{Context, Policy, PolicyResult, Request, Response};
+use http::{HeaderMap, HeaderValue, Method};
+use ring::hmac;
+use std::sync::Arc;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SharedKeyAuthorizationPolicy {
+    base_url: String,
+    shared_key: String,
+    storage_account_name: String,
+}
+
+impl SharedKeyAuthorizationPolicy {
+    #[allow(unused)]
+    pub(crate) fn new(base_url: String, shared_key: String, storage_account_name: String) -> Self {
+        Self {
+            base_url,
+            shared_key,
+            storage_account_name,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Policy for SharedKeyAuthorizationPolicy {
+    async fn send(
+        &self,
+        ctx: &Context,
+        request: &mut Request,
+        next: &[Arc<dyn Policy>],
+    ) -> PolicyResult<Response> {
+        if next.is_empty() {
+            return Err(Box::new(azure_core::PipelineError::InvalidTailPolicy(
+                "Authorization policies cannot be the last policy of a pipeline".to_owned(),
+            )));
+        }
+
+        let headers_mut = request.headers_mut();
+        headers_mut.append(
+            azure_core::headers::MS_DATE,
+            HeaderValue::from_str(
+                format!("{}", chrono::Utc::now().format("%a, %d %h %Y %T GMT")).as_str(),
+            )?,
+        );
+        headers_mut.append(
+            azure_core::headers::VERSION,
+            HeaderValue::from_str("2019-12-12")?,
+        ); // TODO: Remove duplication with storage_account_client.rs
+
+        let uri_path = &request.uri().path_and_query().unwrap().to_string()[1..];
```

Comment:
Do we want to panic when path_and_query is `None`?

Reply:
Probably not :) Code lifted from the Cosmos auth policy, so same problem there: https://github.com/Azure/azure-sdk-for-rust/blob/ad4805942985772e2d3f1adbc08a45eb16d7a48a/sdk/cosmos/src/authorization_policy.rs#L57